xarray/core/parallelcompat.py

```diff
@@ -0,0 +1,272 @@
+    Detects which parallel backend should be used for given set of arrays.
+
+    Also checks that all arrays are of same chunking type (i.e. not a mix of cubed and dask).
+    """
+
+    # TODO this list is probably redundant with something inside xarray.apply_ufunc
+    ALLOWED_NON_CHUNKED_TYPES = {int, float, np.ndarray}
+
+    chunked_arrays = [
+        a
+        for a in args
+        if is_chunked_array(a) and type(a) not in ALLOWED_NON_CHUNKED_TYPES
+    ]
+
+    # Asserts all arrays are the same type (or numpy etc.)
+    chunked_array_types = {type(a) for a in chunked_arrays}
+    if len(chunked_array_types) > 1:
+        raise TypeError(
+            f"Mixing chunked array types is not supported, but received multiple types: {chunked_array_types}"
+        )
+    elif len(chunked_array_types) == 0:
+        raise TypeError("Expected a chunked array but none were found")
+
+    # iterate over defined chunk managers, seeing if each recognises this array type
+    chunked_arr = chunked_arrays[0]
+    chunkmanagers = list_chunkmanagers()
+    for chunkmanager in chunkmanagers.values():
+        if chunkmanager.is_chunked_array(chunked_arr):
+            return chunkmanager
+
+    raise TypeError(
+        f"Could not find a Chunk Manager which recognises type {type(chunked_arr)}"
+    )
+
+
+class ChunkManagerEntrypoint(ABC, Generic[T_ChunkedArray]):
+    """
+    Adapter between a particular parallel computing framework and xarray.
+
+    Attributes
+    ----------
+    array_cls
+        Type of the array class this parallel computing framework provides.
+
+        Parallel frameworks need to provide an array class that supports the array API standard.
+        Used for type checking.
+    """
+
+    array_cls: type[T_ChunkedArray]
+    available: bool = True
+
+    @abstractmethod
+    def __init__(self):
+        ...
+
+    def is_chunked_array(self, data: Any) -> bool:
+        return isinstance(data, self.array_cls)
+
+    @abstractmethod
+    def chunks(self, data: T_ChunkedArray) -> T_NormalizedChunks:
+        ...
+
+    @abstractmethod
+    def normalize_chunks(
+        self,
+        chunks: T_Chunks,
+        shape: tuple[int] | None = None,
+        limit: int | None = None,
+        dtype: np.dtype | None = None,
+        previous_chunks: T_NormalizedChunks | None = None,
+    ) -> T_NormalizedChunks:
+        """Called by open_dataset"""
+        ...
+
+    @abstractmethod
+    def from_array(
+        self, data: np.ndarray, chunks: T_Chunks, **kwargs
+    ) -> T_ChunkedArray:
+        """Called when .chunk is called on an xarray object that is not already chunked."""
+        ...
+
+    def rechunk(
+        self,
+        data: T_ChunkedArray,
+        chunks: T_NormalizedChunks | tuple[int, ...] | T_Chunks,
+        **kwargs,
+    ) -> T_ChunkedArray:
+        """Called when .chunk is called on an xarray object that is already chunked."""
+        return data.rechunk(chunks, **kwargs)  # type: ignore[attr-defined]
```

Comment:
Because if I remove it I get

```
xarray/core/parallelcompat.py:197: error: "T_ChunkedArray" has no attribute "rechunk"  [attr-defined]
```
To specify that the type `T_ChunkedArray` has a `.rechunk` method, I would need a Protocol for `T_ChunkedArray`. ASAIK we don't even have a protocol use to use for non-chunked arrays yet, so I thought I should just ignore this error and leave for later.